File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/main/block/BlockData.java
package pize.tests.voxelgame.main.block;

import pize.tests.voxelgame.client.block.BlockProperties;
import pize.tests.voxelgame.client.block.Blocks;

public class BlockData{
    
    public static BlockProperties getProps(short data){
        return Blocks.fromID(getID(data));
    }
    
    public static BlockProperties getProps(byte ID){
        return Blocks.fromID(ID);
    }

    public static short getData(BlockProperties block, byte extraData){
        return getData(block.getID(), extraData);
    }

    public static short getData(int id, byte extraData){
        return (short) ((id & 0xFF | (((short) extraData) << 8)));
    }

    public static short getData(BlockProperties block){
        return getData(block.getID());
    }
    
    public static short getData(int id){
        return getData(id, (byte) 0);
    }
    
    
    public static byte getID(short data){
        return (byte) (data & 0xFF);
    }
    
    public static byte getState(short data){
        return (byte) (data >> 8);
    }

}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/main/block/BlockData.kt
package pize.tests.voxelgame.main.block

import pize.tests.voxelgame.client.block.BlockProperties
import pize.tests.voxelgame.client.block.Blocks

object BlockData {
    fun getProps(data: Short): BlockProperties? {
        return Blocks.fromID(getID(data))
    }

    fun getProps(ID: Byte): BlockProperties? {
        return Blocks.fromID(ID)
    }

    fun getData(block: BlockProperties?, extraData: Byte): Short {
        return getData(block.getID().toInt(), extraData)
    }

    fun getData(id: Int, extraData: Byte): Short {
        return (id and 0xFF or (extraData.toShort().toInt() shl 8)).toShort()
    }

    fun getData(block: BlockProperties): Short {
        return getData(block.id.toInt())
    }

    fun getData(id: Int): Short {
        return getData(id, 0.toByte())
    }

    fun getID(data: Short): Byte {
        return (data.toInt() and 0xFF).toByte()
    }

    fun getState(data: Short): Byte {
        return (data.toInt() shr 8).toByte()
    }
}
